[12] 일반회원 : 상품 주문

ohsuha opened the pull request.

ohsuha committed:
[12] 상품 주문

f-lab-kai-c commented on src/main/java/org/example/commerce_site/application/order/OrderDetailService.java:
```diff
@@ -0,0 +1,25 @@
+
+	@Transactional
+	public List<OrderDetail> createOrderDetails(List<OrderRequestDto.CreateDetail> details, Order order) {
+		return details.stream()
+			.map(dto -> orderDetailRepository.save(OrderRequestDto.CreateDetail.toEntity(dto, order)))
+			.toList();
```
save를 bulk로 처리하는 방법은 없을까요?

ohsuha replied:
bulk insert 관련해서 찾아보니
1. saveAll() 로 처리하는 방법 : 하지만 사실은 save 쿼리가 여러번 돌아 성능이 많이 좋아지진 않는다.
2. jdbc에서 batchUpdate 를 하는 방법
3. Native Query를 사용
이렇게 세가지 방법이 있는 것같네요 2번이나 3번 방향으로 수정하도록 하겠습니다!

f-lab-kai-c commented on src/main/java/org/example/commerce_site/application/order/OrderFacade.java:
```diff
@@ -0,0 +1,35 @@
+	private final ShipmentService shipmentService;
+	private final AddressService addressService;
+
+	@Transactional
```
기존에 Facde 레이어에서 트랜잭션을 관리하는 것은 거의 없었던 것 같은데, 여기서는 왜 트랜잭션을 통해 관리해야 겠다고 생각했나요?

ohsuha replied:
보낸 요청이 하나의 트랜잭션으로 묶여야할때 사용하는데, 해당 로직에서는 각 서비스 별로 수행되는 로직이 하나의 트랜잭션으로 묶여야 하기 때문이라고 생각했습니다. order detail 데이터를 저장 성공하더라도 다음순서인 order 데이터 저장시 실패하면 order detail 도 롤백 되어야 한다고 생각했기 때문입니다

f-lab-kai-c commented on src/main/java/org/example/commerce_site/application/shipment/ShipmentService.java:
```diff
@@ -0,0 +1,32 @@
+
+	@Transactional
+	public void createShipment(Order order, List<OrderDetail> orderDetail, Address address) {
+		for (OrderDetail orderDetailItem : orderDetail) {
```
for문 사용과 stream 사용에는 어떤 차이가 있을까요?

ohsuha replied:
제가 알고 있었던건 작은 데이터의 처리에는 for가 stream에 비해 오버헤드가 적어서 작은 데이터 셋일 경우 조금 더 빠르다는 것을 보아서 for를 사용했는데, 더 찾아보니 stream 으로 처리하면 기존 컬렉션이나 배열의 값을 변경하지 않고 새로운 값을 생성하는 방식으로 불변성을 유지할 수 있다는 것을 알게 되었습니다..!

f-lab-kai-c replied:
데이터가 큰 셋인 경우에는 stream이 더 빠른 성능을 보이나요?

ohsuha replied:
저는 for가 그동안 더 오래 써와서 최적화가 잘 되어있다고 하는것을 듣고 for를 더 선호해왔는데 관련해서 이런 글을 찾아봤습니다 for가 일반적으로 더 빠르다고 알고 있었는데 참조형의 경우에는 크게 차이가 없군요..!
https://sigridjin.medium.com/java-stream-api%EB%8A%94-%EC%99%9C-for-loop%EB%B3%B4%EB%8B%A4-%EB%8A%90%EB%A6%B4%EA%B9%8C-50dec4b9974b

f-lab-kai-c commented on src/main/java/org/example/commerce_site/attribute/OrderStatus.java:
```diff
@@ -0,0 +1,22 @@
+		this.phase = phase;
+	}
+
+	public boolean isPhaseAtLeast(int value) {
+		return this.phase >= value;
+	}
```
어떤 상황을 위해 존재하는 함수인가요? 사용 사례가 잘 상상되지 않아서요

ohsuha replied:
나중에 사용자가 주문을 취소할때 phase 를 통해 발송 이전 단계에서만 취소를 할때 체크하려고 미리 만들어 두었습니다

f-lab-kai-c replied:
취소 가능한 상태를 return 하는 함수와 위의 함수가 있을 때, 어떤 것이 가독성이 더 좋을까요?

ohsuha replied:
취소 가능한 상태를 return 하는 쪽이 더 좋을것같네요 수정하도록 하겠습니다!

f-lab-kai-c commented on src/main/java/org/example/commerce_site/domain/Order.java:
```diff
@@ -0,0 +1,33 @@
+public class Order extends BaseTimeEntity {
+	private Long userId;
+	private BigDecimal totalAmount;
+	@Enumerated(EnumType.STRING)
```
enum을 string으로 저장하는 것과 number로 저장하는 건 어떤 차이가 있을까요?

ohsuha replied:
스트링으로 저장하면 enum 의 이름 그대로 저장되어 가독성이 좋아지고 중간에 다른 단계를 추가해도 그대로 사용할 수 있습니다. 숫자로 저장하게 되면 저장공간이 줄어들고 인덱스 성능이 좋아지는 걸로 알고 있습니다..!

ohsuha committed:
[12] order detail 을 batchUpdate 를 통해 bulk insert

ohsuha committed:
[12] 배송지 입력도 batchUpdate 를 사용하도록 수정

ohsuha committed:
[11] 카트 수량 수정시 수량 0 체크를 먼저 하도록 수정

ohsuha committed:
[11] ~in 메서드를 통해 한번에 조회하도록 수정

ohsuha committed:
[11] uri 자원들을 복수형 표현으로 수정

ohsuha committed:
[11] 불필요한 파라미터 삭제

- amount price 삭제

ohsuha committed:
[12] cart 의 userid, product id 를 unique 로 설정 후 인덱스 설정

f-lab-kai-c commented on src/main/java/org/example/commerce_site/application/order/dto/OrderDetailResponseDto.java:
```diff
@@ -0,0 +1,32 @@
+public class OrderDetailResponseDto {
+	@Builder
+	public static class Get {
+		protected LocalDateTime createdAt;
```
왜 이 필드는 `protected` 인가요? 그리고 이 값은 API 통해 전달될 때, 어떤 포맷으로 전달되나요?

ohsuha replied:
헉 private 으로 되어야하는데; 실수입니다; yyyy-MM-dd HH-mm-ss 형식으로 전달됩니다..!

f-lab-kai-c commented on src/main/java/org/example/commerce_site/infrastructure/order/OrderDetailBulkRepository.java:
```diff
@@ -0,0 +1,51 @@
+
+	@Transactional
+	public List<OrderDetailResponseDto.Get> saveAll(List<OrderDetail> orderDetails, Long orderId) {
+		String sql = "INSERT INTO order_details (order_id, product_id, quantity, unit_price, created_at) "
+			+ "VALUES (?, ?, ?, ?, ?)";
+
+		jdbcTemplate.batchUpdate(sql, orderDetails, orderDetails.size(),
+			(PreparedStatement ps, OrderDetail orderDetail) -> {
+				ps.setLong(1, orderDetail.getOrder().getId());
+				ps.setLong(2, orderDetail.getProductId());
+				ps.setLong(3, orderDetail.getQuantity());
+				ps.setBigDecimal(4, orderDetail.getUnitPrice());
+				ps.setTimestamp(5, Timestamp.valueOf(orderDetail.getCreatedAt()));
+			});
+
+		String selectSql = "SELECT id, product_id, quantity, unit_price, created_at FROM order_details WHERE order_id = ?";
+
+		List<OrderDetailResponseDto.Get> updatedOrderDetails = jdbcTemplate.query(
+			selectSql,
+			new Object[] {orderId},
+			(rs, rowNum) ->
+				OrderDetailResponseDto.Get.builder()
+					.id(rs.getLong("id"))
+					.productId(rs.getLong("product_id"))
+					.quantity(rs.getLong("quantity"))
+					.unitPrice(rs.getBigDecimal("unit_price"))
+					.orderId(orderId)
+					.createdAt(rs.getTimestamp("created_at").toLocalDateTime())
+					.build()
+		);
+		return updatedOrderDetails;
```
엄밀하게 따지면, insert 했던 것들을 가져오는 것이 보장되지 않는데 repository 레벨에서 두 쿼리를 묶는 것이 좋을까요?

ohsuha replied:
두 쿼리를 분리하는 방식으로 수정했습니다!

f-lab-kai-c commented on src/main/resources/db/migration/mysql/V1.3__set_unique_key_and_add_index_cart.sql:
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE ecommerce_site.carts ADD CONSTRAINT carts_unique UNIQUE KEY (user_id,product_id);
+CREATE INDEX carts_user_id_IDX USING BTREE ON ecommerce_site.carts (user_id,product_id);
```
인덱스를 걸 때, (user_id, product_id) 순으로 거는 것과 (product_id, user_id) 순으로 거는 것에는 어떤 차이가 있을까요?

인덱스는 DB에서 취급하는 정보지만, 어플리케이션 레이어에서 인덱스 정보를 알고 있어야 효율적인 로직을 작성할 수 있습니다. 지금 구조에서는 특정 테이블에 어떤 인덱스가 존재하는 지 확인하려면, flyway 파일에서 정보를 찾아보거나 db에 접속해서 정보를 확인해야합니다. 이를 개선할 방법이 있을까요?

ohsuha replied:
1. 다중 컬럼 인덱스 순서에 따라 쿼리 성능에 차이가 생길 수 있습니다. user_id 가 먼저이면 user_id 인덱스가 우선적으로 정렬되어있기때문에 user_id 로 검색하고 반대라면 먼저 product_id 가 일치하는 것부터 먼저 비교 연산이 실행됩니다. 동일한 product_id 를 여러 유저가 장바구니에 담을 수 있고, 장바구니는 user_id 에 따라 보여야하기 때문에 user_id 가 먼저인게 맞는것같습니다.
2. 테이블에 @Table (indexes = @Index(name = "idx_column", columnList = "column_name")) 와같이 설정한 인덱스를 표기하는 방법이 있을것같습니다.

ohsuha committed:
[12] 코드리뷰 내용 수정

- repository 단위의 트랜잭션 분리 (bulk insert 후 읽어오기)
- OrderStatus 의 주문 취소 상태 판단 함수 수정
- protected 로 잘못 설정되어있던 변수들 private 으로 수정

f-lab-kai-c commented on src/main/java/org/example/commerce_site/infrastructure/order/OrderDetailBulkRepository.java:
```diff
@@ -0,0 +1,51 @@
+				ps.setTimestamp(5, Timestamp.valueOf(orderDetail.getCreatedAt()));
+			});
+
+		// String selectSql = "SELECT id, product_id, quantity, unit_price, created_at FROM order_details WHERE order_id = ?";
+		//
+		// List<OrderDetailResponseDto.Get> updatedOrderDetails = jdbcTemplate.query(
+		// 	selectSql,
+		// 	new Object[] {orderId},
+		// 	(rs, rowNum) ->
+		// 		OrderDetailResponseDto.Get.builder()
+		// 			.id(rs.getLong("id"))
+		// 			.productId(rs.getLong("product_id"))
+		// 			.quantity(rs.getLong("quantity"))
+		// 			.unitPrice(rs.getBigDecimal("unit_price"))
+		// 			.orderId(orderId)
+		// 			.createdAt(rs.getTimestamp("created_at").toLocalDateTime())
+		// 			.build()
+		// );
+		// return updatedOrderDetails;
```
제거 안하고 주석으로 남겨놓는 이유가 있을까요?

ohsuha committed:
[12] 코드리뷰 반영

- 불필요한 주석제거
- 테이블에 index 표기
